### PR TITLE
Modularize DLC Interface using companion objects

### PR DESCRIPTION
**Pull request by nkohen**
Before this, all DLC functionality required direct access to `DLCMessage`s meaning all functions constructed state from "backups" which is inefficient, hard to follow, and hard to test.

This will make testing modules easier (future PR) and many calls can be made more efficiently in the wallet (future PR).

Note: PSBTs will be further removed in a future PR.

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/DLCMessage.scala [hunk 1]
```diff
@@ -23,6 +23,12 @@ object DLCMessage {
       .flip
   }
 
+  def computeContractId(
+      fundingTx: Transaction,
+      tempContractId: Sha256Digest): ByteVector = {
+    fundingTx.txIdBE.bytes.xor(tempContractId.bytes)
+  }
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/DLCMessage.scala]: This should probably be in `DLCUtil`

File: dlc/src/main/scala/org/bitcoins/dlc/sign/DLCTxSigner.scala [hunk 2]
```diff
@@ -71,299 +72,125 @@ case class DLCTxSigner(
...
-      pubKey: ECPublicKey): Future[PartialSignature] = {
-    val sigOpt = psbt.inputMaps.head.partialSignatures
-      .find(_.pubKey == pubKey)
-
-    sigOpt match {
-      case None =>
-        Future.failed(new RuntimeException("No signature found after signing"))
-      case Some(partialSig) => Future.successful(partialSig)
-    }
+  def completeFundingTx(remoteSigs: FundingSignatures): Future[Transaction] = {
```
Comment (benthecarman) [inline dlc/src/main/scala/org/bitcoins/dlc/sign/DLCTxSigner.scala:80]: Would be nice to have a 

```scala
def completeFundingTx(localSigs: FundingSignatures, remoteSigs: FundingSignatures): Transaction
```

Comment (benthecarman) [reply dlc/src/main/scala/org/bitcoins/dlc/sign/DLCTxSigner.scala:80]: same for cet

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/DLCUtil.scala [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      sigPubKey
+        .extractAdaptorSecret(adaptorSig, ECDigitalSignature(cetSig.bytes.init))
+        .fieldElement
+
+    SchnorrDigitalSignature(outcome.aggregateNonce, possibleOracleS)
+  }
+
+  def computeOutcome(
+      completedSig: ECDigitalSignature,
+      possibleAdaptorSigs: Vector[(OracleOutcome, ECAdaptorSignature)]): (
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/DLCUtil.scala]: nit: Maybe just return `Option[(SchnorrDigitalSignature,OracleOutcome)]`? Seems like it would be useful to signify if it was found or not with an `Option[]` rather than throwing

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/DLCUtil.scala [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    }
+  }
+
+  def computeOutcome(
+      isInitiator: Boolean,
+      offerFundingKey: ECPublicKey,
+      acceptFundingKey: ECPublicKey,
+      contractInfo: ContractInfo,
+      localAdaptorSigs: Vector[(OracleOutcome, ECAdaptorSignature)],
+      cet: Transaction): (SchnorrDigitalSignature, OracleOutcome) = {
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/DLCUtil.scala]: any reason we don't just tighten this type up to be `cet: WitnessTransaction`

File: dlc/src/main/scala/org/bitcoins/dlc/builder/DLCTxBuilder.scala [hunk 5]
```diff
@@ -84,88 +84,307 @@ case class DLCTxBuilder(offer: DLCOffer, accept: DLCAcceptWithoutSigs)(implicit
...
+    val offerTotalFunding: CurrencyUnit =
+      offerFundingInputs.map(_.output.value).sum
+
+    // The sum of all funding input amounts from the non-initiator
+    val acceptTotalFunding: CurrencyUnit =
+      acceptFundingInputs.map(_.output.value).sum
+
+    require(
+      offerTotalFunding >= offerInput,
+      "Offer funding inputs must add up to at least offer's total collateral")
```
Comment (Christewart) [inline dlc/src/main/scala/org/bitcoins/dlc/builder/DLCTxBuilder.scala]: nits: Add the amounts to the error message, always helps when debugging :-)

```scala
 require(
      offerTotalFunding >= offerInput,
      "Offer funding inputs must add up to at least offer's total collateral, offerTotalFunding=$offerTotalFunding offerInput=$offerInput")
```

Comment (Christewart) [reply dlc/src/main/scala/org/bitcoins/dlc/builder/DLCTxBuilder.scala]: same with below

File: dlc/src/main/scala/org/bitcoins/dlc/builder/DLCTxBuilder.scala [hunk 6]
```diff
@@ -84,88 +84,307 @@ case class DLCTxBuilder(offer: DLCOffer, accept: DLCAcceptWithoutSigs)(implicit
...
+  def buildCETs(
+      outcomes: Vector[OracleOutcome],
+      contractInfo: ContractInfo,
+      offerFundingKey: ECPublicKey,
+      offerFinalSPK: ScriptPubKey,
+      acceptFundingKey: ECPublicKey,
+      acceptFinalSPK: ScriptPubKey,
+      timeouts: DLCTimeouts,
+      fundingOutputRef: OutputReference): Vector[
+    (OracleOutcome, WitnessTransaction)] = {
```
Comment (Christewart) [inline dlc/src/main/scala/org/bitcoins/dlc/builder/DLCTxBuilder.scala]: Seems like it would be worthwhile to make a

```scala
/** An oracle outcome and it's corresponding CET */
case class OutcomeCETPair(outcome: OracleOutcome, wtx: WitnessTransaction)
```

File: dlc/src/main/scala/org/bitcoins/dlc/execution/DLCExecutor.scala [hunk 7]
```diff
@@ -85,33 +85,74 @@ case class DLCExecutor(signer: DLCTxSigner)(implicit ec: ExecutionContext) {
...
+  def executeRefundDLC(dlcSetup: SetupDLC): RefundDLCOutcome = {
+    val SetupDLC(fundingTx, _, refundTx) = dlcSetup
+    RefundDLCOutcome(fundingTx, refundTx)
+  }
+}
+
+object DLCExecutor {
 
-    val threshold = dlcSetup.cets.head._1.oracles.length
+  def executeDLC(
```
Comment (Christewart) [inline dlc/src/main/scala/org/bitcoins/dlc/execution/DLCExecutor.scala:116]: Could you add a scaladoc on this method for what the behavior is when we find a signature that does not correspond to an outcome? 

I have a hard time understanding what is happening in these scenarios, as we should be able to always recover the outcome from the signature -- at least that is my understanding?

Specifically i'm talking about https://github.com/nkohen/bitcoin-s-core/blob/ac4aa8edd6063e726e54567a0c72e93a8be74c89/dlc/src/main/scala/org/bitcoins/dlc/execution/DLCExecutor.scala#L132

File: dlc/src/main/scala/org/bitcoins/dlc/execution/DLCExecutor.scala [hunk 8]
```diff
@@ -85,33 +85,74 @@ case class DLCExecutor(signer: DLCTxSigner)(implicit ec: ExecutionContext) {
...
-        (msg, cetInfo.remoteSignature)
+
+    val msgAndCETInfoOpt = msgOpt.flatMap { msg =>
+      remoteCETInfos.find(_._1 == msg)
+    }
+
+    val (msg, ucet, remoteAdaptorSig) = msgAndCETInfoOpt match {
+      case Some((msg, CETInfo(ucet, remoteSig))) => (msg, ucet, remoteSig)
       case None =>
         throw new IllegalArgumentException(
```
Comment (Christewart) [inline dlc/src/main/scala/org/bitcoins/dlc/execution/DLCExecutor.scala:140]: so is this a "fatal" error? This seems really bad to me, an outcome occurred we didn't know about?

File: dlc/src/main/scala/org/bitcoins/dlc/sign/DLCTxSigner.scala (hunk 2 above)
Comment (Christewart) [reply dlc/src/main/scala/org/bitcoins/dlc/sign/DLCTxSigner.scala:80]: I'm guessing this is needed because of current `Sign` API, nadav is working on making a synchronous version in #2652

**Review by Christewart — CHANGES_REQUESTED** [review]

File: dlc/src/main/scala/org/bitcoins/dlc/execution/DLCExecutor.scala (hunk 8 above)
Comment (nkohen) [reply dlc/src/main/scala/org/bitcoins/dlc/execution/DLCExecutor.scala:140]: I mean if all the oracles sign "hehe, screw my users" there isn't really anything you can do about that.

**Review by Christewart — APPROVED** [review]

**Review by benthecarman — APPROVED** [review]
